k8s/secrets.go

```diff
@@ -23,25 +19,89 @@ func GetSecret(ctx context.Context, name, namespaceOverride string) (*corev1.Sec
 	if err != nil {
 		return
 	}
 
-	client, err := NewKubernetesClientset(namespace)
+	secrets, err := client.CoreV1().Secrets(namespace).List(ctx, metav1.ListOptions{})
 	if err != nil {
 		return
 	}
 
-	secrets, err := client.CoreV1().Secrets(namespace).List(ctx, metav1.ListOptions{})
+	for _, s := range secrets.Items {
+		names = append(names, s.Name)
+	}
+
+	return
+}
+
+func CreateDockerRegistrySecret(ctx context.Context, name, namespaceOverride, username, password, server string) (err error) {
+	client, namespace, err := NewClientAndResolvedNamespace(namespaceOverride)
 	if err != nil {
 		return
 	}
 
-	for _, s := range secrets.Items {
-		names = append(names, s.Name)
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Type: corev1.SecretTypeDockerConfigJson,
+		Data: map[string][]byte{},
 	}
 
+	dockerConfigJSONContent, err := handleDockerCfgJSONContent(username, password, "", server)
+	if err != nil {
+		return
+	}
+	secret.Data[corev1.DockerConfigJsonKey] = dockerConfigJSONContent
+
+	_, err = client.CoreV1().Secrets(namespace).Create(ctx, secret, metav1.CreateOptions{})
 	return
 }
+
+// --- Helper methods for DockerConfigJson type of Secret
+// Taken from (and converted to private):
+// https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/create/create_secret_docker.go
+
+// DockerConfigJSON represents a local docker auth config file
+// for pulling images.
+type dockerConfigJSON struct {
+	Auths dockerConfig `json:"auths" datapolicy:"token"`
+	// +optional
+	HttpHeaders map[string]string `json:"HttpHeaders,omitempty" datapolicy:"token"`
+}
+
+// DockerConfig represents the config file used by the docker CLI.
+// This config that represents the credentials that should be used
+// when pulling images from specific image repositories.
+type dockerConfig map[string]dockerConfigEntry
+
+// dockerConfigEntry holds the user information that grant the access to docker registry
+type dockerConfigEntry struct {
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty" datapolicy:"password"`
+	Email    string `json:"email,omitempty"`
+	Auth     string `json:"auth,omitempty" datapolicy:"token"`
+}
+
+func handleDockerCfgJSONContent(username, password, email, server string) ([]byte, error) {
+	dockerConfigAuth := dockerConfigEntry{
+		Username: username,
+		Password: password,
+		Email:    email,
+		Auth:     encodeDockerConfigFieldAuth(username, password),
+	}
+	dockerConfigJSON := dockerConfigJSON{
+		Auths: map[string]dockerConfigEntry{server: dockerConfigAuth},
+	}
+
+	return json.Marshal(dockerConfigJSON)
+}
+
+// encodeDockerConfigFieldAuth returns base64 encoding of the username and password string
+func encodeDockerConfigFieldAuth(username, password string) string {
+	fieldValue := username + ":" + password
+	return base64.StdEncoding.EncodeToString([]byte(fieldValue))
```

Comment:
I'm probably just missing it somewhere, but I didn't see where this is ever used/decoded.

Reply:
It's just a few lines above in the `handleDockerCfgJSONContent()` func.